### PR TITLE
move fork-able, checkpoint-able state into BankState

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem

a slice of Entries arriving from the window stage can start at an arbitrary
  Entry, current single-linked list of state checkpoints Bank can't deal
  with that

  #### Summary of Changes

introduce an inner "BankState", to allow Bank to have multiple "heads" along with a HashMap of read-only checkpoints that are parents of those heads

Fixes #

### CODE REVIEW

File: src/bank.rs [hunk 1]
```diff
@@ -146,147 +134,536 @@ pub struct LastIdEntry {
...
 /// Manager for the state of all accounts and contracts after processing its entries.
 pub struct Bank {
-    /// A map of account public keys to the balance in that account.
-    pub accounts: RwLock<Accounts>,
+    /// current bank state and all checkpoints
+    pub heads: RwLock<VecDeque<RwLock<BankState>>>,
 
-    /// FIFO queue of `last_id` items
-    last_ids: RwLock<LastIds>,
+    checkpoints: RwLock<HashMap<Hash, BankState>>,
```
Comment (rob-solana) [inline src/bank.rs]: this is the interesting part, but has turned into a huge mess, where everything's now implemented by BankState...

**Comment by rob-solana** [comment]
cc #1783, which is more incremental approach

**Comment by carllin** [comment]
It might be a good idea to get the design down in writing and then have a design meeting about this, it's getting quite complicated.

**Comment by aeyakovenko** [comment]
@rob-solana can we move the "execute/process" part of the bank out into a separate file?

**Comment by rob-solana** [comment]
this change is outdated after most recent discussion on implementation, which makes keeping multiple forks live the problem of the window/ledger/rocksdb blob manager

yes: there could be a bank_engine or bank_runtime or something, but it's not necessary right now except perhaps for cleanliness

**Pull request closed without merging**